discord/http.py

```diff
@@ -349,159 +349,170 @@ class Ratelimit:
 
     This is similar to a semaphore except tailored to Discord's rate limits. This is aware of
     the expiry of a token window, along with the number of tokens available. The goal of this
-    design is to increase throughput of requests being sent concurrently rather than forcing
-    everything into a single lock queue per route.
+    design is to increase throughput of requests being sent concurrently by forcing everything
+    into a single lock queue per route.
     """
 
     __slots__ = (
+        'key',
         'limit',
         'remaining',
+        'reset_at',
+        'pending',
         'outgoing',
-        'reset_after',
-        'expires',
-        'dirty',
+        'one_shot',
+        'http',
         '_last_request',
-        '_max_ratelimit_timeout',
-        '_loop',
-        '_pending_requests',
-        '_sleeping',
+        '_lock',
+        '_future',
     )
 
-    def __init__(self, max_ratelimit_timeout: Optional[float]) -> None:
+    def __init__(self, http: HTTPClient, key: str) -> None:
+        self.key = key
         self.limit: int = 1
         self.remaining: int = self.limit
+        self.reset_at: float = 0.0
+        self.pending: int = 0
         self.outgoing: int = 0
-        self.reset_after: float = 0.0
-        self.expires: Optional[float] = None
-        self.dirty: bool = False
-        self._max_ratelimit_timeout: Optional[float] = max_ratelimit_timeout
-        self._loop: asyncio.AbstractEventLoop = asyncio.get_running_loop()
-        self._pending_requests: deque[asyncio.Future[Any]] = deque()
-        # Only a single rate limit object should be sleeping at a time.
-        # The object that is sleeping is ultimately responsible for freeing the semaphore
-        # for the requests currently pending.
-        self._sleeping: asyncio.Lock = asyncio.Lock()
-        self._last_request: float = self._loop.time()
+        self.one_shot: bool = False
+        self.http: HTTPClient = http
+        self._last_request: float = 0.0
+        self._lock: asyncio.Lock = asyncio.Lock()
+        self._future: Optional[asyncio.Future] = None
```

Comment:
By moving this to a `Future` you're basically saying you expect a single waiter.

```suggestion
        self._future: Optional[asyncio.Future[None]] = None
```